File: backend/routers/bundles.py
import logging
import traceback
from fastapi import APIRouter, Depends, HTTPException, BackgroundTasks, UploadFile, File
from sqlalchemy.ext.asyncio import AsyncSession
from typing import List

from ..database import get_db
from ..models.bundle import Bundle
from ..services.bundle_service import BundleService
from ..schemas.common import ApiResponse, ApiStatus
import shutil
from pathlib import Path

logger = logging.getLogger("api.bundles")

router = APIRouter(
    prefix="/bundles",
    tags=["Bundles"]
)

@router.get("/", response_model=ApiResponse[List])
async def list_bundles(db: AsyncSession = Depends(get_db)):
    """List all available workflow bundles."""
    try:
        from sqlalchemy.future import select
        result = await db.execute(select(Bundle))
        bundles = result.scalars().all()
        # Serialize to dicts to avoid ORM serialization issues
        data = [
            {
                "id": b.id,
                "name": b.name,
                "version": b.version,
                "author": b.author,
                "description": b.description,
                "tags": b.tags or [],
                "is_installed": b.is_installed,
                "installed_workflow_id": b.installed_workflow_id,
                "thumbnail_url": b.thumbnail_url,
                "created_at": str(b.created_at),
                "updated_at": str(b.updated_at),
            }
            for b in bundles
        ]
        return ApiResponse(status=ApiStatus.SUCCESS, data=data)
    except Exception as e:
        logger.error(f"Error listing bundles: {e}\n{traceback.format_exc()}")
        raise HTTPException(status_code=500, detail=f"Internal server error: {str(e)}")

@router.post("/import")
async def import_bundle(file: UploadFile = File(...), db: AsyncSession = Depends(get_db)):
    """Upload a .zip workflow bundle."""
    if not file.filename.endswith('.zip'):
        raise HTTPException(status_code=400, detail="Must be a .zip file")

    try:
        service = BundleService(db)
        temp_path = service.imports_dir / file.filename
        
        with open(temp_path, "wb") as buffer:
            shutil.copyfileobj(file.file, buffer)

        bundle = await service.import_bundle(str(temp_path))
        return ApiResponse(status=ApiStatus.SUCCESS, data={"status": "success", "bundle_id": bundle.id})
    except HTTPException:
        raise
    except Exception as e:
        if 'temp_path' in locals() and Path(temp_path).exists():
            Path(temp_path).unlink()
        logger.error(f"Error importing bundle: {e}\n{traceback.format_exc()}")
        raise HTTPException(status_code=400, detail=str(e))

@router.post("/{workflow_id}/export")
async def export_workflow(workflow_id: str, db: AsyncSession = Depends(get_db)):
    """Export an existing workflow into a .zip bundle."""
    try:
        service = BundleService(db)
        zip_path = await service.export_bundle(workflow_id)
        return ApiResponse(status=ApiStatus.SUCCESS, data={"status": "success", "file_path": zip_path})
    except Exception as e:
        logger.error(f"Error exporting bundle {workflow_id}: {e}\n{traceback.format_exc()}")
        raise HTTPException(status_code=400, detail=str(e))

@router.post("/{bundle_id}/install")
async def install_bundle(bundle_id: str, db: AsyncSession = Depends(get_db)):
    """Install a bundle, creating its workflow entry."""
    try:
        service = BundleService(db)
        workflow = await service.install_bundle(bundle_id)
        return ApiResponse(status=ApiStatus.SUCCESS, data={"workflow_id": workflow.id if workflow else None})
    except Exception as e:
        logger.error(f"Error installing bundle {bundle_id}: {e}\n{traceback.format_exc()}")
        raise HTTPException(status_code=400, detail=str(e))

@router.delete("/{bundle_id}")
async def delete_bundle(bundle_id: str, db: AsyncSession = Depends(get_db)):
    """Delete a bundle from the registry."""
    try:
        from sqlalchemy.future import select
        result = await db.execute(select(Bundle).where(Bundle.id == bundle_id))
        bundle = result.scalar_one_or_none()
        if not bundle:
            raise HTTPException(status_code=404, detail="Bundle not found")
        await db.delete(bundle)
        await db.commit()
        return ApiResponse(status=ApiStatus.SUCCESS, data={"deleted": True})
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error deleting bundle {bundle_id}: {e}\n{traceback.format_exc()}")
        raise HTTPException(status_code=500, detail=str(e))

File: frontend/lib/api/bundles.ts
import { api, ApiResponse } from './client';
import { AIAsset } from './assets';

export interface BundleDependency {
  asset_type: string;
  name: string;
  is_optional: boolean;
}

export interface BundleManifest {
  name: string;
  version: string;
  author: string;
  description: string;
  tags: string[];
  required_assets: BundleDependency[];
}

export interface Bundle {
  id: string;
  name: string;
  version: string;
  author: string;
  description: string;
  tags: string[];
  required_assets: BundleDependency[];
  installed_workflow_id: string | null;
  is_installed: boolean;
  thumbnail_url: string | null;
  created_at: string;
  updated_at: string;
}

export interface BundleCompatibilityReport {
  is_compatible: boolean;
  missing_required: BundleDependency[];
  missing_recommended: BundleDependency[];
  found_assets: AIAsset[];
}

export interface BundleImportResponse {
  status: string;
  bundle_id: string;
}

export interface BundleExportResponse {
  status: string;
  file_path: string;
}

export const BundleService = {
  getBundles: async (): Promise<ApiResponse<Bundle[]>> => {
    // Note: Assuming backend returns List[Bundle] without wrapping based on our router implementation
    return api.get<Bundle[]>('/bundles/');
  },

  getBundle: async (id: string): Promise<ApiResponse<Bundle>> => {
    return api.get<Bundle>(`/bundles/${id}`);
  },

  importBundle: async (file: File): Promise<ApiResponse<BundleImportResponse>> => {
    return api.upload<BundleImportResponse>('/bundles/import', file, {});
  },

  exportBundle: async (workflowId: string): Promise<ApiResponse<BundleExportResponse>> => {
    return api.post<BundleExportResponse>(`/bundles/${workflowId}/export`, {});
  },

  installBundle: async (id: string): Promise<ApiResponse<{ workflow_id: string }>> => {
    return api.post<{ workflow_id: string }>(`/bundles/${id}/install`, {});
  },
  
  deleteBundle: async (id: string): Promise<ApiResponse<null>> => {
    return api.delete<null>(`/bundles/${id}`);
  }
};

// -------------------------------------------------------------------------
// Compatibility Engine Utilities
// -------------------------------------------------------------------------
export const checkBundleCompatibility = (bundle: Bundle, availableAssets: AIAsset[]): BundleCompatibilityReport => {
  const missing_required: BundleDependency[] = [];
  const missing_recommended: BundleDependency[] = [];
  const found_assets: AIAsset[] = [];

  const availableAssetNames = new Set(availableAssets.map(a => a.filename));

  bundle.required_assets.forEach(dep => {
    // Very simple matcher: checks if the exact filename exists in the registry
    if (availableAssetNames.has(dep.name)) {
      const asset = availableAssets.find(a => a.filename === dep.name);
      if (asset) found_assets.push(asset);
    } else {
      if (dep.is_optional) {
        missing_recommended.push(dep);
      } else {
        missing_required.push(dep);
      }
    }
  });

  return {
    is_compatible: missing_required.length === 0,
    missing_required,
    missing_recommended,
    found_assets
  };
};
